measureme/src/counters.rs

```diff
@@ -0,0 +1,869 @@
+
+            Ok(counter)
+        }
+
+        /// Try to access the mmap page, retrying the `attempt` closure as long
+        /// as the "seqlock" sequence number changes (which indicates the kernel
+        /// has updated one or more fields within the mmap page).
+        #[inline]
+        fn access_mmap_page_with_seqlock<T>(
+            &self,
+            attempt: impl Fn(&perf_event_mmap_page) -> T,
+        ) -> T {
+            // FIXME(eddyb) it's probably UB to use regular reads, especially
+            // from behind `&T`, with the only synchronization being barriers.
+            // Probably needs atomic reads, and stronger ones at that, for the
+            // `lock` field, than the fields (which would be `Relaxed`?).
+            let mmap_page = unsafe { &*(self.mmap.as_ptr() as *const perf_event_mmap_page) };
+            let barrier = || std::sync::atomic::fence(std::sync::atomic::Ordering::Acquire);
+
+            loop {
+                // Grab the "seqlock" - the kernel will update this value when it
+                // updates any of the other fields that may be read in `attempt`.
+                let seq_lock = mmap_page.lock;
+                barrier();
+
+                let result = attempt(mmap_page);
+
+                // If nothing has changed, we're done. Otherwise, keep retrying.
+                barrier();
+                if mmap_page.lock == seq_lock {
+                    return result;
+                }
+            }
+        }
+    }
+
+    impl super::HwCounterRead for Counter {
+        type Output = u64;
+
+        #[inline]
+        fn read(&self) -> u64 {
+            // HACK(eddyb) keep the accurate code around while not using it,
+            // to minimize overhead without losing the more complex implementation.
+            let (counter, offset, pmc_width) = if cfg!(accurate_seqlock_rdpmc) && false {
+                self.access_mmap_page_with_seqlock(|mp| {
+                    let caps = unsafe { mp.__bindgen_anon_1.__bindgen_anon_1 };
+                    assert_ne!(caps.cap_user_rdpmc(), 0);
+
+                    (
+                        rdpmc(mp.index.checked_sub(1).unwrap()),
+                        mp.offset,
+                        mp.pmc_width,
+                    )
+                })
+            } else {
+                (rdpmc(self.reg_idx), 0, 48)
+            };
+
+            let counter = offset + (counter as i64);
+
+            // Sign-extend the `pmc_width`-bit value to `i64`.
+            (counter << (64 - pmc_width) >> (64 - pmc_width)) as u64
+        }
+    }
+
+    impl super::HwCounterRead for (&Counter, &Counter) {
+        type Output = (u64, u64);
+
+        #[inline]
+        fn read(&self) -> (u64, u64) {
+            // HACK(eddyb) keep the accurate code around while not using it,
+            // to minimize overhead without losing the more complex implementation.
+            if (cfg!(accurate_seqlock_rdpmc) || cfg!(unserialized_rdpmc)) && false {
```

Comment:
I think we should document these cfg flags somewhere, perhaps in the module docs.

Reply:
The `|| true` (or `&& false`, here) are so that they're not actually exposed, see also https://github.com/rust-lang/measureme/pull/143#discussion_r522040961